jaeger-core/src/main/java/io/jaegertracing/internal/clock/SystemClock.java

```diff
@@ -14,19 +14,45 @@
 
 package io.jaegertracing.internal.clock;
 
+import lombok.val;
+
 /**
- * Default implementation of a clock that delegates its calls to the system clock. The
- * microsecond-precision time is simulated by (millis * 1000), therefore the
- * {@link #isMicrosAccurate()} is false.
+ * This implementation of the system-clock will provide true microseconds accurate timestamp,
+ * given that the JVM supports it (JDK 9 and above).
+ * <p>
+ * The actual timestamp generation implementation for both scenarios can be found in {@link CurrentTimeSupplier}.
  *
- * @see System#currentTimeMillis()
+ * @author <a href="mailto:ishinberg0@gmail.com">Idan Sheinberg</a>
+ * @see CurrentTimeSupplier#micros()
  * @see System#nanoTime()
  */
 public class SystemClock implements Clock {
 
+  private static final boolean MICROS_ACCURATE;
+
+  private static final CurrentTimeSupplier CURRENT_TIME_SUPPLIER;
+
+  private static int getJavaVersion() {
+    val sections = System.getProperty("java.version").split("\\.");
+    val major = Integer.parseInt(sections[0]);
+    return major == 1 ? Integer.parseInt(sections[1]) : major;
+  }
+
+  private static CurrentTimeSupplier getCurrentTimeSupplier() {
+    return MICROS_ACCURATE
+        ? CurrentTimeSupplier.MicrosAccuracy.INSTANCE
+        : CurrentTimeSupplier.MillisAccuracy.INSTANCE;
+  }
+
+  static {
+    val version = getJavaVersion();
+    MICROS_ACCURATE = version >= 9;
```

Comment:
What about 
```java
try {
  Class.forName("java.lang.StackWalker");
  MICROS_ACCURATE = true;
} catch( ClassNotFoundException e ) {
  MICROS_ACCURATE = false;
}
```

Reply:
Implict assumption of the JDK version by the existence of a class (which is more prone to changes) as opposed to literally and explicitly checking the condition we wish to validate - the actual JDK version, is less than optimal, and at least in my opinion, a poor practice.

As this is not my code-base, I'd do whatever the maintainer wishes, but I vote neigh.